visidata/memory.py

```diff
@@ -32,4 +32,4 @@ def memosSheet(vd):
 
 
 Sheet.addCommand('Alt+Shift+M', 'open-memos', 'vd.push(vd.memosSheet)', 'open the Memory Sheet')
-Sheet.addCommand('Alt+m', 'memo-cell', r'vd.memory[input("assign \""+cursorCol.getDisplayValue(cursorRow)+"\" to: ")] = cursorCol.getTypedValue(cursorRow)', 'store value in current cell in Memory Sheet')
+Sheet.addCommand('Alt+m', 'memo-cell', r'w = min(vd.options.default_width, windowWidth-len("assign \"\" to: ")-1)-1; dv = cursorCol.getDisplayValue(cursorRow); dv = dv[:w-dispwidth(options.disp_truncator)] + options.disp_truncator if dispwidth(dv) > w else dv[:w]; vd.memory[input("assign \""+dv+"\" to: ")] = cursorCol.getTypedValue(cursorRow)', 'store value in current cell in Memory Sheet')
```

Comment:
This seems pretty complicated..what about using `inputMultiple` and putting the value into the secondary field?  Like `assign: x+2` and `to: `.

Reply:
That is much cleaner:
```
test_input_cmd = 'vd.memory[inputMultiple(name=dict(prompt="to memo name: "), cell=dict(prompt="assign cell value: "+cursorCol.getDisplayValue(cursorRow),display=False))["name"]] = cursorCol.getTypedValue(cursorRow);'
Sheet.addCommand('V', 'test-input', test_input_cmd, 'tester')
```
But it triggers the same error if the user hits Tab to switch to the long prompt showing the cell value. Maybe if we add an argument `inputMultiple()` to leave off the bindings passed to `input()`?
https://github.com/saulpw/visidata/blob/7ad1de0b98fb6b77c786e0198cae49167c391c6e/visidata/_input.py#L468

Reply:
Oh, I mean use the value as the initial value of the "assign cell value" field (`value=`).  Then tabbing to it will allow you to edit the value, too, and `inputMultiple` should already be able to handle that (or we should fix it).

Also I believe you can use `cursorTypedValue` instead of `cursorCol.getTypedValue(cursorRow)`.

And when the execstr gets to be this size, let's move some of it into function(s), to make it more readable (I know I am guilty of this too more than anyone :)

Reply:
I tried making the value editable, but it seemed like the edited value of `inputMultiple()` was a string.  (because I was editing a string from `getDisplayValue()`). That's why I displayed the value in the prompt, to avoid giving the user a chance to change the type.

Now I tried applying the column type to the, copying `editCell()`: https://github.com/saulpw/visidata/blob/5e611e39cbb6a35c25bd34c323956f9a7bd44c30/visidata/_input.py#L639

```
def memo_cell(sheet):
    value_params = dict(prompt='assign value: ',
                        value=sheet.cursorCol.getDisplayValue(sheet.cursorRow))
    name_params = dict(prompt='to memo name: ')
    inputs = vd.inputMultiple(memo_name=name_params, memo_value=value_params)
    try:
        v_edited = sheet.cursorCol.type(inputs['memo_value'])  # convert input to column type
    except ValueError:
        vd.warning('the new memo value\'s type does not match the column type')
        v_edited = inputs['memo_value']
    if not inputs['memo_name']:
        vd.fail('memo name cannot be blank')
    vd.memory[inputs['memo_name']] = v_edited
Sheet.addCommand('V', 'test-input', 'memo_cell(vd.sheet)', 'test variation of memo-cell')
```

But a cell with a list inside, like `[1,2,3]` in a JSON sheet, turns into a string `'[3] 1; 2; 3'`, even if the user never tabs into it. Is there a better way to proceed?

Also, now that I've moved the code into a function, is `cursorDisplay` still preferred over `getDisplayValue()`? Or is it only preferred in `addCommand` strings, for readability?

Reply:
`cursorDisplay` is preferred, because it can be cached.

Reply:
> But a cell with a list inside, like [1,2,3] in a JSON sheet, turns into a string '[3] 1; 2; 3', even if the user never tabs into it. Is there a better way to proceed?

Oh, now I understand.  That's not great.  One idea is, we could add a `readonly` or something kwarg to `input()`, and make it so that readonly prompts can't be edited, and thus won't be accepted.

Another possibility would be to add a special case for lists and dicts to parse them correctly after input in general.  I've run into cases myself where I want to edit the value of a cell with `e` and have to use `z=` instead, which is non-intuitive.  But this would be a bigger change, because this display format isn't compatible with any existing parser, and I don't think we should be committing to this display format which is a little janky anyway.  So maybe we just go with the easier first option for now?

Reply:
Okay, I went with the first option, making the field readonly, inside `inputMultiple()`. It can't be tabbed into to be edited.